src/protocols/echo/CHIPEchoServer.cpp

```diff
@@ -0,0 +1,89 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements an object for a CHIP Echo unsolicitied
+ *      responder (server).
+ *
+ */
+
+#include "CHIPEcho.h"
+
+namespace chip {
+namespace Protocols {
+
+ChipEchoServer::ChipEchoServer()
+{
+    ExchangeMgr           = nullptr;
+    OnEchoRequestReceived = nullptr;
+}
+
+CHIP_ERROR ChipEchoServer::Init(ExchangeManager * exchangeMgr)
+{
+    // Error if already initialized.
+    if (ExchangeMgr != nullptr)
+        return CHIP_ERROR_INCORRECT_STATE;
+
+    ExchangeMgr           = exchangeMgr;
+    OnEchoRequestReceived = nullptr;
+
+    // Register to receive unsolicited Echo Request messages from the exchange manager.
+    ExchangeMgr->RegisterUnsolicitedMessageHandler(kChipProtocol_Echo, kEchoMessageType_EchoRequest, HandleEchoRequest, this);
```

Comment:
would love to see us model this with delegates instead

Reply:
The delegate pattern does not apply here since this pattern only works when the callback function's is fixed. The unsolicited message handler varies from protocol to protocol and this handler might also get updated within the same protocol depending on the context.

Reply:
I see, thanks.